csharp/src/Google.Protobuf/Reflection/MessageDescriptor.cs

```diff
@@ -104,6 +104,8 @@ internal MessageDescriptor(DescriptorProto proto, FileDescriptor file, MessageDe
             {
                 map[JsonFormatter.ToCamelCase(field.Name)] = field;
                 map[field.Name] = field;
+                if (!string.IsNullOrEmpty(field.JsonName))
```

Comment:
We may also not need the `map[JsonFormatter.ToCamelCase]` assignment here. I believe the intention is that protoc will *always* populate the json_name part of the descriptor, either converting the case or using the specified option. We should keep the use of the original name though.